### ISSUE TITLE
When connection to db refuses error is not triggered

### ISSUE THREAD

**Issue by mrbar42** [arangodb/arangojs#184]
The call to the database is this:

``` javascript
db.collection('list')
    .byExample({type: 'content'})
    .then(cursor => cursor.all())
    .catch(error => {
      console.error("ERROR", error);
    });
```

The request fails with an uncaught `TypeError`
The error is due to unreachable database e.g. `connection refused`
this is the error with its stack 

```
Uncaught TypeError: Cannot read property 'content-type' of undefined
    (anonymous function)   @ connection.js?7a47:117
    callback               @ request.js?e840:90
    (anonymous function)   @ request.js?e840:98
    EventEmitter.emit      @ events.js?7c71:74
    Response.handle        @ response.js?57b6:99
    xhr.onreadystatechange @ request.js?a434:73
```

Which comes from this line in `connection.js`

``` javascript
if (res.headers['content-type'].match(MIME_JSON)) {
```

Investigating this a bit showed that in the case of connection refused error,
the `end` event from response object is emitted with zero length body
before the `error` event from request object is emitted - there for the request is treated as success.

I tried all kind of things,
and this snippet seems to overcome the wrong ordering of events but still allows the end to fire eventualy.

changing this part in `utils/request.js`

``` javascript
.on('end', () => {
          res.body = data.join('');
          callback(null, res);
        });
```

to this - and by that allowing the error to propagate first if it exists

``` javascript
.on('end', () => {
          res.body = data.join('');
          if (!res.body) return setImmediate(()=>_callback(null, res))
          callback(null, res);
        });
```

Per haps there is a better place to handle this case - for example the response status in such case is 0.

**bottom line**
When a connection is refused the order of things is as follow:
1.  request fails - readystate changes
2. `end` emitted     (`data` never gets called)
3.  callback invoked without error (TypeError here)
4. `error` emitted with `Network error`
5.  callback invoked with the error - ignored

**Comment by pluma** [comment]
Interesting. I was fairly certain the error handler would be invoked before the end callback in this situation.

I'll have to investigate. Sorry for the delay.

**Comment by pluma** [comment]
I can't seem to reproduce this error anymore. This probably means it has been fixed at some point. Can you try again with the development version or the alpha release for 6.0 (`npm install arangojs@6.0.0-alpha`)? Feel free to reopen this issue if the error still exists.
